### PR TITLE
port the cluster e2e to use the framework

### PR DESCRIPTION
**Pull request by stevekuznetsov**
config: allow bootstrapping other CRDs

Signed-off-by: Steve Kuznetsov <skuznets@redhat.com>

---

cluster: improve logging

Signed-off-by: Steve Kuznetsov <skuznets@redhat.com>

---

e2e: framework: allow exposing raw config

For encoding the raw kubeconfig for our servers, we need to be able to
access it directly.

Signed-off-by: Steve Kuznetsov <skuznets@redhat.com>

---

e2e: framework: allow naming kcp servers

For better UX when poking around inside of the artifact directory for
tests with more than one kcp server, we allow naming them.

Signed-off-by: Steve Kuznetsov <skuznets@redhat.com>

---

e2e: workspace: scope testCase variable

Without this scoping step, the reference leaks out to each test
goroutine.

Signed-off-by: Steve Kuznetsov <skuznets@redhat.com>

---

hack: generate e2e code as well

Signed-off-by: Steve Kuznetsov <skuznets@redhat.com>

---

makefile: add a helper for running e2e tests

Since the e2e runs kcp servers via the binary, we need to make sure that
a) the built binaries are in $PATH and b) we re-build when running.

Signed-off-by: Steve Kuznetsov <skuznets@redhat.com>

---

e2e: cluster: write the test suite

Signed-off-by: Steve Kuznetsov <skuznets@redhat.com>

---

e2e: generate code

Signed-off-by: Steve Kuznetsov <skuznets@redhat.com>

---

hack: remove old e2e test

Signed-off-by: Steve Kuznetsov <skuznets@redhat.com>

---

requires #212

### CODE REVIEW

**Comment by stevekuznetsov** [comment]
@sttts @ncdc there's a lot here between the test-specific CRD+client stuff, the bespoke logic to create e.g the `Cluster` and wait for it, etc - what (unfortunately) was a few lines of `kubectl` turns into a lot of Go. We gain a lot for it, of course, but ...

**Comment by stevekuznetsov** [comment]
Failing due to #269 - also please let me know if you want me to break the commits up into separate pulls.

**Comment by stevekuznetsov** [comment]
Flaked again on #286

**Comment by stevekuznetsov** [comment]
Rebased on top of #289

**Comment by sttts** [comment]
needs another rebase

File: config/bootstrap.go [hunk 1]
```diff
@@ -67,12 +67,18 @@ func BootstrapCustomResourceDefinitions(ctx context.Context, client apiextension
...
+// BootstrapCustomResourceDefinitionFromFS creates the CRD using the target client from the
+// provided filesystem handle and waits for it to become established. This call is blocking.
+func BootstrapCustomResourceDefinitionFromFS(ctx context.Context, client apiextensionsv1client.CustomResourceDefinitionInterface, gk metav1.GroupKind, fs embed.FS) error {
 	start := time.Now()
 	klog.Infof("bootstrapping %v", gk.String())
 	defer func() {
 		klog.Infof("bootstrapped %v after %s", gk.String(), time.Since(start).String())
 	}()
-	raw, err := rawCustomResourceDefinitions.ReadFile(fmt.Sprintf("%s_%s.yaml", gk.Group, gk.Kind))
+	raw, err := fs.ReadFile(fmt.Sprintf("%s_%s.yaml", gk.Group, gk.Kind))
```
Comment (sttts) [inline config/bootstrap.go:81]: does this change anything beyond the refactoring?

File: pkg/reconciler/cluster/cluster.go [hunk 2]
```diff
@@ -172,7 +172,7 @@ func (c *Controller) reconcile(ctx context.Context, cluster *clusterv1alpha1.Clu
 				"SyncerInstalling",
 				"Installing syncer on cluster")
 		case SyncerModeNone:
-			klog.Info("syncer ready!")
+			klog.Info("started none more syncer!")
```
Comment (sttts) [inline pkg/reconciler/cluster/cluster.go:175]: none more?

**Review by sttts — APPROVED** [review]

File: pkg/reconciler/cluster/cluster.go (hunk 2 above)
Comment (stevekuznetsov) [reply pkg/reconciler/cluster/cluster.go:175]: `none mode` * :(

File: config/bootstrap.go (hunk 1 above)
Comment (stevekuznetsov) [reply config/bootstrap.go:81]: No - it's just factoring out the filesystem.

**Comment by stevekuznetsov** [comment]
@sttts rebased, fixed the logging typo

**Comment by stevekuznetsov** [comment]
Hm, have not seen *that* flake before - can't reproduce locally even with `-count 100`. Added a commit to improve the error message.

**Comment by stevekuznetsov** [comment]
...finally

**Pull request merged**